### PR TITLE
Support properties

### PR DESCRIPTION
**Pull request by the-jos**

### CODE REVIEW

File: autobump-maven/src/main/java/com/github/autobump/maven/model/MavenDependencyResolver.java [hunk 1]
```diff
@@ -17,23 +18,44 @@
 
     @Override
     public Set<Dependency> resolve(Workspace workspace) {
+        Model model = getModel(workspace);
+        return model
```
Comment (Gschrooyen) [inline autobump-maven/src/main/java/com/github/autobump/maven/model/MavenDependencyResolver.java:22]: ik zou dit op 1 lijn zetten in plaats van 2
`return getModel(workspace)`...

File: autobump-maven/src/test/java/com/github/autobump/maven/model/MavenDependencyResolverTest.java [hunk 2]
```diff
@@ -29,18 +29,51 @@ public void TestSuccesresolve() {
...
+                        .group("org.apache.derby")
+                        .name("derby")
+                        .version("10.15.2.0")
+                        .build()),
+                deps);
+    }
+
+    @Test
+    public void TestresolveNullProperty() {
+        assertThrows(DependencyParserException.class, () ->
```
Comment (timpeeters) [inline autobump-maven/src/test/java/com/github/autobump/maven/model/MavenDependencyResolverTest.java:54]: I would argue that if a dependency is defined using a version property: ${...} and we cannot find the property, that it is best for us to ignore the dependency instead of throwing an exception and stopping the entire dependency resolution.

There could be cases where the developer uses ${project.version} for example to define the version of the dependency. The project.version will not be found in the <properties/> section but it is still perfectly valid.

In Maven, it is also possible to provide property values from the command line. E.g. devs could run: `mvn package -Dorg.apache.derby.version=1.0`. Again, the property is not defined in the `pom.xml` file but it is still valid.

So bottom line, I think it's safer to ignore unresolved properties instead of throwing exceptions.

File: autobump-maven/src/main/java/com/github/autobump/maven/model/MavenDependencyResolver.java [hunk 3]
```diff
@@ -17,23 +18,44 @@
...
+        }
+        String version = model.getProperties().getProperty(getPropertyName(dependencyVersionData));
+        if (version == null) {
+            throw new DependencyParserException("Error reading version property");
+        }
+        return version;
+    }
 
+    private boolean isProperty(String dependencyVersionData) {
+        return dependencyVersionData.startsWith("${") && dependencyVersionData.endsWith("}");
```
Comment (timpeeters) [inline autobump-maven/src/main/java/com/github/autobump/maven/model/MavenDependencyResolver.java:55]: This is typically something we solve by using regular expressions.

```
private static final Pattern VERSION_PROPERTY_PATTERN = Pattern.compile("\\$\\{(.+)}");

Matcher matcher = VERSION_PROPERTY_PATTERN.matcher(...)

if (matcher.matches()) {
    matcher.group(1); // Here we extract group 1 from the match (.+), the value between ${ and }
}
```

File: autobump-maven/src/main/java/com/github/autobump/maven/model/MavenDependencyResolver.java (hunk 1 above)
Comment (the-jos) [reply autobump-maven/src/main/java/com/github/autobump/maven/model/MavenDependencyResolver.java:22]:
```suggestion
 return getModel(workspace)
                .getDependencies()
                .stream()
                .filter(dependency -> dependency.getVersion() != null)
                .map(dependency -> Dependency.builder()
                        .group(dependency.getGroupId())
                        .name(dependency.getArtifactId())
                        .version(getDependencyVersionFromModel(getModel(workspace), dependency.getVersion()))
                        .build())
                .collect(Collectors.toSet());
```

Dan moet ik wel 2x getModel doen... Tim is het beter om het dan toch zo te laten? Of haal ik het 2x op?

File: autobump-maven/src/test/java/com/github/autobump/maven/model/MavenDependencyResolverTest.java [hunk 4]
```diff
@@ -50,11 +50,16 @@ public void TestSuccesresolveProperties() {
...
-    public void TestresolveNullProperty() {
-        assertThrows(DependencyParserException.class, () ->
-                dependencyResolver.resolve(
-                        new Workspace("src/test/resources/project_root_support_properties_nullproperty")
-                ));
+    public void TestresolveUndevinedProperty() {
+        Workspace ws = new Workspace("src/test/resources/project_root_support_properties_undefinedproperty");
+        Set<Dependency> deps = dependencyResolver.resolve(ws);
+        assertEquals(
+                Set.of(Dependency.builder()
```
Comment (timpeeters) [inline autobump-maven/src/test/java/com/github/autobump/maven/model/MavenDependencyResolverTest.java:57]: I think the assertion here should be that the set of dependencies is empty. Otherwise, Autobump will start looking for this dependency in the VersionRepository using version: ${org.apache.derby.version}

Unresolved version property -> Autobump ignores the dependency.

Comment (LanderTuyteleersXplore) [reply autobump-maven/src/test/java/com/github/autobump/maven/model/MavenDependencyResolverTest.java]: @the-jos Ik denk dat er hier bedoeld wordt om de versions die niet geresolved worden door bv. een malformed of undefined version niet op te nemen in het resolven van de dependencies. Momenteel worden die dependencies mee ingelezen. 

`    @Test
    public void TestresolveMalformedVarProperty() {
        Workspace ws = new Workspace("src/test/resources/project_root_support_properties_badproperty");
        Set<Dependency> deps = dependencyResolver.resolve(ws);
        assertEquals(
                Set.of(Dependency.builder()
                        .group(TEST_DEPENDENCY_GROUP)
                        .name(TEST_DEPENDENCY_NAME)
                        .version("${org.apache.derby.version")
                        .build()),
                deps);
    }`

Deze test zou dan een lege lijst moeten teruggeven in plaats van de malformed version.
Correct me if I'm wrong @timpeeters
